docs/settings.md
# Settings

The `settings.yml` file is divided into three main sections

    transports:
        ...

    recipients:
        ...

    watchers:
        ....

## Transports

Transports are used to send alerts when watcher states change. 

### SMTP Transport 

You can send email using any SMTP server. AWD? uses simple user/password SMTP authenticaiton, and has been confirmed working with Amazon SES and MailGun. For example, to set up Mailgun as a transport, use

    transports:
        smtp:
            server : smtp.mailgun.org
            port : 587
            secure : false
            user : postmaster@<your-mailgun-account>.mailgun.org
            pass: your-mailgin-password
            from : me@example.com

With that out of the way, it should be said that SMTP is really simple to set up if you're doing everything
exactly right, and difficult to diagnose the instant something goes wrong. If you're struggling with settings 
to work, first test them independently with a simple SMTP client like `swaks`:

        sudo apt install swaks -y

        ./swaks --auth \
            --server smtp.some-service.com \
            --au postmaster@YOUR_DOMAIN_NAME \
            --ap your-password \
            --to bar@example.com \
            --h-Subject: "Hello world" \
            --body 'Testing send you some mail'

If that works, AWD? has its own STMP test that sends your a mail directly. Use it with

        curl http://yourAWDserver/diagnostics/transport/smtp/<recipient-node>

where `recipient-node` is the recipient node in AWD's config with an `smtp:email-address` child property on it.

### Slack Transport

Alerts can be sent to a Slack user or channel using a custom app that acts as a bot - check out [apps dashboard](https://api.slack.com/apps). 

- Your Slack app needs four scopes, `channels:read, groups:read, mpim:read, im:read, chat:write`. To post to channels you also need to manually add your  app to the channel in question, via the Slack app.
- Secret is also called "Signing Secret" is specific to your Slack app.
- Token normally starts with `xox` and is specific to your app's integration into your workspace. To find the token, open "add features and functionality" then click on "permissions", then look for "Bot User OAuth Token".

        transports:
            slack:
                secret: ...............
                token: xoxb-...........

## Recipients

Recipients are people or systems that receive alerts. 

### SMTP

To receive emails add `smtp` to a recipient and add email address to it

    recipients:
        BobMcName:
            smtp: bob@example.com

### Slack 

To receive Slack alerts add `slack` to a recipient (either a person or something imaginary for channel alerts) and define either a Slack userid or channelid.

    recipients:

        BobMcName:
            slack: <BobMcName's slack userId>

        foo:
            slack: <slack channelId>

To get a user id, click on the user profile in the Slack desktop client, then look under "More". To find a channel id, open Slack in a browser and click on the channel you want to post to - the channel id is the second id in browser address bar https://app.slack.com/client/workspace-id/channel-id/user_profile/your-own-user-id

## Watchers

Watchers let you test something at regular intervals. The following built-in tests are available.

### HTTP 200

The simplest (and default) watcher is the HTTP check. It queries a host and fails if it doesn't get an HTTP code 200 back. You can specify another HTTP code if you expect something other than 200 (such as a `403` if the host prompts you to login).

    watchers:
        mytest:
            host: example.com

            # optional code if you're expecting something other than 2xx
            code : 403 

### Port open

Test if a port at the given host is open. Works on TCP only.

    watchers:
        port:
            test: net.portOpen
            host: 192.168.1.126
            port: 8006

### Jenkins job status

Test if a jenkins job is passing. Requires a Jenkins server URL and job name.

- Host can be any URL that gives access to the server, this is often with built-in credentials. 
- Job name can be the human-friendly version, we'll make it URL-safe, so copy this directly from your Jenkins UI if you want.
- By default, a test passes on success only, all other outcomes will be read as failure. You can override this with `status` value, which can be a comma-separated string consisting of `success`, `aborted` and/or `failure` (these strings are taken directly from the Jenkins API).

        watchers:
            my_jenkins_job:
                test : jenkins.buildSuccess
                host: <USER>:<PASSWORD>@example.com
                job: My Jenkins Job

                # Optional.
                status: success,aborted 

### Docker container up

If your have the Docker [HTTP API](https://docs.docker.com/engine/api/v1.24/) enabled, you can query it to test if a container is up on a given docker host. 

    watchers:
        my-container-test:
            test: docker.containerIsUp
            host: example.com
            container: myContainer # container name

            # Optional. Port to query, the default is 2375.
            port: 2375 

### System.d service running

You can test if a system.d service is running. You will need SSH access to the machine running the service. Password can be templated in via an env var ([see advanced settings](#Secrets)).

    watchers:
        my-service-test:
            test: systemd.servicerunning
            host: example.com
            user: myuser
            password: mypassword

            # name of the system.d service to query
            service: docker

### Disk use 

You can check if a disk is runnnig out of space (Linux only). You will need SSH access to the machine running the service. Password can be templated in via an env var ([see advanced settings](#Secrets)).

    watchers:
        my-disk-use-test:
            test: resource.linux.diskUse
            host: example.com
            user: myuser
            password: mypassword
            path: /some/path/on/host

            # max % use allowed
            threshold: 50

### Ping

You can ping a host. 

- Timeout is in seconds, and is optional. 

        watchers:
            my-ping-test:
                test: net.ping
                host: example.com

                # optional
                timeout: 10 

### ZFS Pool Status

You can query the status of a ZFS pool on a host. This test ensures that the pool's `state` is `online`. You will need SSH access to the machine running the service. Password can be templated in via an env var ([see advanced settings](#Secrets)).

    watchers:
        is-my-data-gone:
            test: filesystem.zfs.zpoolStatus
            host: example.com
            user: myuser
            password: mypassword

            # name of zfs pool to check
            pool: mypoolname
                
### Date In File

You can rig up a simple "dead man's switch" by writing a date to a file at the end of some process. As long as the date in the file is current, the test will pass. The date written must be in Javascript-parsable ISO format. You can generate a file in a Linux terminal with the command

    echo $(date --iso-8601=seconds) >> /path/to/datefile

You will need SSH access to the machine running the service. Password can be templated in via an env var ([see advanced settings](#Secrets)).

    watchers:
        is-my-process-alive:
            test: general.dateInFile
            host: example.com
            user: myuser
            password: mypassword

            # path on remote machine date is written to
            path: /path/to/datefile

            # Maximum allowed age of date in file. Can be any digit followed by S, M, H or D (seconds, minutes, hours or days, case not important)
            range: 24H

### Default watcher behaviour

- Watchers have a default interval of 1 minute. You can override any watcher's default with any valid cronmask (masks must be in quotes, this is a YML quirk).
- The YML node is the default name of a watcher. You can provide a display name using the optional `name` field.
- Watchers, like most other objects in *AreWeDown?* have an optional `enabled` field that defaults to true. Set this to false to disable the watcher.
- To alert specific people about a watcher failure, use `recipients` - this is an optional, comma-separated list of names defined under the top-level `recipients` section. If left empty, all recipients will receive alerts for that watcher.

        watchers:
            mytest:
                
                # Give it a fancier name with spaces and things
                name: my "fancy" test name!

                # Send alerts to these people only. Sane spacing optional (Python trigger warning)
                recipients: BobMcName,someOtherPerson,      YetAnotherPerson            

                # Run it on a Tuesday only, because.
                interval: '0 0 * * TUE'
                                
                # Actually, don't run it at all.
                enabled: false

## Custom tests

You can write your own tests in any shell script supported by your host system. See [custom tests](/custom-tests.md) for more.

## Advanced Settings

### Defaults

The following overridable default settings live in the root-level of settings.yml.

    # default dashboard title 
    header: Are We Down?

    # defines where logs are written. Default value is relative to application startup path. If you want to 
    # write to /var/log/arewedown for example, change this
    logs: ./logs

    # Port HTTP server run ons   
    port: 3000

    # Interval for dashboard refreshes (milliseconds)
    dashboardRefreshInterval: 5000,

    # Interval for dashboard timeout (milliseconds)
    dashboardLoadTimeout: 5000,

    # Allows app to be restarted from user interface
    UIRestart: false

    # Amount of data logged. Can be error, warn, info
    logLevel: warn

    # internal work cleans up/maintains self. needs to run once a day only. Must be wrapped in quotes.
    internalWorkerTimer : '0 0 * * *'

    # In days. AreWeDown? can clean up its own log data to prevent your disk from flooding.
    logRetention: 365 

### Secrets

If you do not want to store sensitive information like passwords in the `settings.yml` file, you can pass these as environment variables from your host system, and bind them anywhere in settings with the `"{{env.___}}"` template pattern

    anyProperty: "{{env.MY_SENSITIVE_INFO}}"

If you defined an environment variable `MY_SENSITIVE_INFO=1234`, the `anyProperty` setting have the value `1234` in memory. If a templated environment variable is not found, *AreWeDown?* will fail to start.

### Logs

*AreWeDown?* logs, a lot. It writes its own logs to the `/etc/arewedown/logs/<DATE>.log`, and then for each watcher in `/etc/arewedown/logs/<WATCHER>/logs/<DATE>.log`. If *AreWeDown?* fails to start or exits abruptly, a good place to start looking is its logs.


docs/advanced-settings.md
# Advanced Settings

## Defaults

*AreWeDown?* works out of the box with no cutom settings. You can override the defaults with the following. Note that these settings are passed in as standard environment variables, you can define them any way you want. They can be written either at the root level of _settings.yml_ file or in your _docker-compose.yml_'s `environment:` section. 

    // override default title
    header: 'Are We Down?'

    # defines where logs are written. Default value is relative to application startup path
    logs: './logs'

    # Port HTTP server run ons   
    port: 3000

    // Interval dashboard refreshes (milliseconds)
    dashboardRefreshInterval: 5000,

    // Interval for dashboard timeout (milliseconds)
    dashboardLoadTimeout: 5000,

    // Allows app to exit with an HTTP call. This is useful if you're changing configuration in settings.yml and don't want to restart the Docker/daemon process the app runs in
    // If properly daemonized your app will restart instantly and with updated settings applied.
    allowHttpExit: false

    // Amount of data logged. Can be error, warn, info
    logLevel: 'warn'

    // internal work cleans up/maintains self. needs to run once a day only.
    internalWorkerTimer : '0 0 * * *'

    // In days. AreWeDown? can clean up its own log data to prevent your disk from flooding.
    logRetention: 365 

Documentation can sometimes go out of sync with code - check [settings.js](https://github.com/shukriadams/arewedown/blob/master/src/lib/settings.js) to be sure.

## Sensitive data

If you do not want to store sensitive information like passwords in the `settings.yml` file, you can store these as environment variables on your host system, and bind them anywhere in settings with the following template pattern

    anyProperty: "{{env.MY_SENSITIVE_INFO}}"

If you defined an environment variable `MY_SENSITIVE_INFO=1234` then `anyProperty` will have the value `1234` in memory. If the environment variable is not found, *AreWeDown?* will fail to start.

## Logs

*AreWeDown?* logs, a lot. It writes its own logs to the `/etc/arewedown/logs/<DATE>.log`, and then for each watcher in `/etc/arewedown/logs/<WATCHER>/logs/<DATE>.log`. If *AreWeDown?* fails to start or exits abruptly, a good place to start looking is its logs. If you're running in docker, use `docker logs arewdown`.

src/lib/settings.js
let _settings = null

module.exports = {


    /**
     * Explicitly crash app if a value is null or not defined - zero and empty strings must be allowed so don't use
     * falsy.
     */
    failIfNotSet(value, message){
        if (value === null || value === undefined)
            throw message
    },


    /**
     * Checks a setting value for "{{env.<VALUE>}}" content, if found, replaces entire value with the environment variable named <VALUE>. Use this
     * to store sensitive information like passwords in env variables instead of in settings.yml
     * @param {string} value 
     */
    replaceTemplatedEnvVars(value){
        
        const process = require('process'),
            match = value.match(/{{env.(.*)}}/i)
    
        if (match){

            const value = match.pop()
    
            if (process.env[value])
                return process.env[value]
            else 
                throw `ERROR : config expects environment variable "${value}", but this was not found.`
            
        }
    
        return value
    },


    /**
     * 
     */
    searchAndReplaceTemplatedEnvVars(node){
        for (let child in node)
            if (typeof node[child] === 'string')
                node[child] = this.replaceTemplatedEnvVars(node[child])
            // always handle arrays before objects, as arrays are also objects in JS
            else if (Array.isArray(node[child])) 
                for (item of node[child])
                    this.searchAndReplaceTemplatedEnvVars(item)
            else if (typeof node[child] === 'object' ) 
                this.searchAndReplaceTemplatedEnvVars(node[child])
        },


    /**
     * Loads settings from yml file and various env var sources. Also validates settings after load - app can be forced to exit
     * in event of serious misconfiguration.
     * 
     * @param {object} forcedIncomingSettings Used by testing only
     */
    load(forcedIncomingSettings = null){
        let fs = require('fs-extra'),
            process = require('process'),
            yaml = require('js-yaml'),
            dotenv = require('dotenv'),
            sanitize = require('sanitize-filename'),
            disabledWatchersCount = 0,
            allWatcherNames = [],
            disabledRecipients = []

        // apply env vars from optional .env file in project root
        dotenv.config()

        let settingsPath = process.env.AWD_SETTINGS_PATH || './config/settings.yml' // allow settings path to be passed in

        // force wipe for app restart
        _settings = null
        
        if (fs.existsSync(settingsPath)){
            console.log(`Loading settings from ${settingsPath}`)
            try {
                let settingsYML = fs.readFileSync(settingsPath, 'utf8')
                _settings = yaml.safeLoad(settingsYML)
            } catch (e) {
                console.log('Error reading settings.yml', e)
                throw e
            }
        } else {
            console.warn(`!! Warning: settings.yml not found - please create it at "<AreWeDown? root folder>/config/settings.yml".`) 
            console.warn(`!! If you are running in docker, confirm your volume mounts.`)
        }

        // allow forcedIncomingSettings to override all settings load from file
        _settings = forcedIncomingSettings || _settings
        
        // default settings
        _settings = Object.assign({
        
            // override default title
            header: 'Are We Down?',

            // root path
            rootpath: '/',

            // path all log data is written to. This folder is in the AWD? root folder.
            logs : './logs',
            
            /// path transport queue is written too. This directory is in the AWD? root directory.
            queue : './queue',

            // can be error|warn|info|debug in increasing levels of spamminess.
            logLevel: 'warn',

            // port AWD? listens on.
            port: 3000,
            
            // optional string script that can be invoked as shell script when app starts.
            // use this for running custom scripts, installing apt packages etc
            onstart: null,

            //
            allowDevRoutes : false,

            // onstart scripts can sometimes throw warnings that are treated as errors - 
            // set this to true to ignore
            onstartIgnoreError : false,

            // autoassigned protocol if none is given
            defaultTestProtocol : 'http',

            // milliseconds dashboard reloads itself
            dashboardRefreshInterval: 5000,
        
            // milliseconds for dashboard to timeout and show self-error
            dashboardLoadTimeout: 5000,

            // allows AWD to exit (and cleanly restart when daemonized) via HTTP call. Enable only if you trust people
            // on your network not to abuse
            UIRestart: false,

            // internal worker timer, cleans up/self-maintains. Should run once a day only
            internalWorkerTimer : '0 0 * * *',

            // 60 seconds default, interval for sending queued alerts. Pending alerts are grouped into a single message.
            transportWorkerTime: '*/60 * * * * *',

            //
            pagesPerGroup: 10,

            // number of items to display whenever paging is used
            pageSize: 10,

            // 
            listCondenseThresh: 10,

            // in days. set to zero to disable.
            logRetention: 365, 

            // 500 ms by default. 
            // Forced delay, in milliseconds, between consecutive watchers. Under-the-hood, all watchers run on the same cron clock,
            // so jobs that fire on the same regular intervals (every minute, every hour, etc) all fire at exactly the same time.
            // This can cause system bottlenecks when running many watchers. This offset is inserted between consecutive watchers 
            // when starting, forcing them to run staggered.
            watcherOffset: 500,
            
            // root-level objects
            dashboards : {},
            recipients : {},
            transports : {},
            watchers : {},
        
            // DEVELOPMENT STUFF - you will likely never have to change these
        
            // set to false when developing for views to rebuild on each page reload
            cacheViews : true

        }, _settings)
        
        if (!_settings.rootpath.endsWith('/')) 
            _settings.rootpath += '/'
            
        // it's still possible to nullify these with goofy setup.yml content, so force exist as we rely on them being present
        _settings.dashboards = _settings.dashboards || {}
        _settings.recipients = _settings.recipients || {}
        _settings.transports = _settings.transports || {}
        _settings.watchers = _settings.watchers || {}

        // ensure that "default" dashboard is defined if no dashboards set. There must always be at least
        // dashboard in system
        if (!Object.keys(_settings.dashboards).length)
            _settings.dashboards['default'] = { name : '' }
        
        // apply default recipient settings
        for (const recipient in _settings.recipients){
        
            _settings.recipients[recipient] = Object.assign({
                enabled : true
            }, _settings.recipients[recipient])
        
            // remove if disabled
            if (!_settings.recipients[recipient].enabled){
                disabledRecipients.push(recipient)
                delete _settings.recipients[recipient]
            }
        }
        
        // apply default transport settings
        for (const transport in _settings.transports){
            _settings.transports[transport] = Object.assign({

                // set this to false to immediately bypass/disable all communication over this transport
                enabled : true,

                // set to true to allow diagnostic calls from the web API.
                diagnostics: false,

                // set to true to fall back to mock transport provider. Requires that mocking is implemented in the given transport.
                mock: false

            }, _settings.transports[transport])
        
            // remove if disabled
            if (!_settings.transports[transport].enabled)
                delete _settings.transports[transport]
        }

        // allow primative properties on .watchers node to cascade
        // down as defaults on all .watchers child objects
        let watcherGlobalOverrides = {},
            allowedTypes = ['number', 'boolean', 'string']

        for (const name in _settings.watchers){
            const property = _settings.watchers[name]

            if (allowedTypes.includes(typeof property)){
                watcherGlobalOverrides[name] = property
                delete _settings.watchers[name]
            }
        }

        // apply default watcher settings
        for (const watcherId in _settings.watchers){
            let watcher = _settings.watchers[watcherId]

            // apply default watcher settings
            watcher = Object.assign({
                
                // this is the parent node's id repeated
                __id : watcherId,

                // filesystem-friendly version of name
                __safeId : sanitize(watcherId),
                
                // internally set error message. normally for validation text. merged with cronprocess's, .errorMessage
                __errorMessage : null,
        
                // system will flag watchers if they fail start validation. watchers with config errors
                // will not run (to save on error spam in logs), but will still be visible in dashboards
                __hasErrors : false,
        
                // users can add their own convenient name, if not this defaults to node name
                name : watcherId,   
        
                //cronmask to time test - default is 1 minute
                interval : '*/1 * * * *',

                // Slaved to global watcherOffset. Disabled by default.
                offset : 0,

                // internal test to call. must be in src/tests folder, must not have .js extension, egs `net.httpCheck`
                test : null,

                // String of user names to receive alerts on watcher status change. 
                // Will be converted to string array on load.
                // Can be *, in which case everyone in settings.recipients will be pasted in
                recipients : '*',

                // external command. either test or cmd must be given
                cmd : null,
        
                // enabled field is optional and true by default. setting this to false will cause the 
                // watcher to be completely ignored
                enabled : true,
        
            }, watcher)
            
            // apply defaults
            for (let prop in watcherGlobalOverrides)
                if (watcher[prop] === undefined)
                    watcher[prop] = watcherGlobalOverrides[prop]

            // remove if disabled
            if (!watcher.enabled){
                delete _settings.watchers[watcherId]
                disabledWatchersCount ++
                continue
            }
        
            allWatcherNames.push(watcherId)
        
            // ensure user didn't force null on this, from here on we assume this value is set
            watcher.recipients = watcher.recipients || '*' 
        
            // force default values based on logic  
            if (!watcher.cmd && !watcher.test)
                watcher.test = 'net.httpCheck'
        
            _settings.watchers[watcherId] = watcher
        }
        
        // apply default dashboard settings
        for (const dashboardId in _settings.dashboards){
        
            _settings.dashboards[dashboardId] = Object.assign({
                // node id, attached here for convenience
                __id : dashboardId,  
        
                // nodename, made safe for filesystems
                __safeId : sanitize(dashboardId), 
        
                // users can add their own convenient name, if not this defaults to node name
                name : dashboardId,    
        
                enabled : true,
        
                // set to true to have this be the default dashboard when viewing '/' in a browser. if not set, the first
                // dashboard defined will be default. If multiple are defined with default, the first one defined is taken
                default : false,
        
                // force to all watchers
                watchers : '*'  
            }, _settings.dashboards[dashboardId])
        
            // remove if disabled
            if (!_settings.dashboards[dashboardId].enabled){
                delete _settings.dashboards[dashboardId]
                continue
            }
        
            // if dashboard is set to * watchers, replace it's watchers list with literal names of all watchers
            if (_settings.dashboards[dashboardId].watchers.trim() === '*'){
                if (allWatcherNames.length){
                    console.debug(`assigning all watchers to dashboard "${dashboardId}"`)
                    _settings.dashboards[dashboardId].watchers = allWatcherNames.join(',')
                } else {
                    _settings.dashboards[dashboardId].watchers = ''
                }
            }
        }
        
        // if a watcher has no explicit recipients list, assign all recipient names to list
        const allRecipientNames = Object.keys(_settings.recipients)
        for (const watcherId in _settings.watchers){
            const watcher = _settings.watchers[watcherId]
        
            if (watcher.recipients === '*'){
                // replace with array of all recipients, or emty array
                if (allRecipientNames.length){
                    watcher.recipients = allRecipientNames
                    console.debug(`assigning all recipients "${allRecipientNames.join(',')}" to watcher ${watcherId}`)
                } else {
                    watcher.recipients = []
                    console.warn(`WARNING : no default recipients to assign to contactless-watcher ${watcherId} - this watcher will fail silently`)
                }
            } else {
                // convert string list to array if string is not * (if * will be handled later)
                watcher.recipients = watcher.recipients
                    .split(',')
                    .map(w => w ? w.trim() : w)
                    .filter(w => !!w)
            }

            // remove disabled watchers
            watcher.recipients = watcher.recipients.filter(r => !disabledRecipients.includes(r))

            // ensure that recipient names exist in global recipient list
            for (const recipientName of watcher.recipients)
                if (!_settings.recipients[recipientName])
                    throw `Recipient name ${recipientName} in watcher ${watcherId} is not defined under global recipients.`

            // ensure test exists - we cannot verify cmd as that's entirely up to user, and must fail at watcher execute level
            if (watcher.test && !fs.existsSync(`${__dirname}/../tests/${watcher.test}.js`))
                throw `ERROR: watcher "${watcherId}" specifies non-existent test "${watcher.test}".`
        
            _settings.watchers[watcherId] = watcher
        }
        
        // validate SMTP if enabled
        if (_settings.transports.smtp && _settings.transports.smtp.enabled){
            this.failIfNotSet(_settings.transports.smtp.server, 'settings "transports.smtp" is missing expected value ".server"')
            this.failIfNotSet(_settings.transports.smtp.port, 'settings "transports.smtp" is missing expected value ".port"')
            this.failIfNotSet(_settings.transports.smtp.secure, 'settings "transports.smtp" is missing expected value ".secure"')
            this.failIfNotSet(_settings.transports.smtp.user, 'settings "transports.smtp" is missing expected value ".user"')
            this.failIfNotSet(_settings.transports.smtp.pass, 'settings "transports.smtp" is missing expected value ".pass"')
            this.failIfNotSet(_settings.transports.smtp.from, 'settings "transports.smtp" is missing expected value ".from"')
        }

        // validate slack settings if enabled
        if (_settings.transports.slack && _settings.transports.slack.enabled){
            this.failIfNotSet(_settings.transports.slack.token, 'settings "transports.slack" is missing expected value ".token"')
            this.failIfNotSet(_settings.transports.slack.secret, 'settings "transports.slack" is missing expected value ".secret"')
        }

        if (disabledWatchersCount)
            console.warn(`${disabledWatchersCount} watchers disabled`)
        
        // validate watchers
        if (!Object.keys(_settings.watchers).length)
            console.warn('WARNING: No watchers defined or enabled, nothing is being watched and no alerts will be sent')
        
        for (const name in _settings.watchers){
            const watcher = _settings.watchers[name]
        
            // todo : warn on multiple defaults
            this.failIfNotSet(watcher.interval, `Watcher "${name}" has no interval`)
        }
        
         
        this.searchAndReplaceTemplatedEnvVars(_settings)
    },
   

    /**
     * 
     */
    reset(){
        _settings = null
    },


    get(){
        if (!_settings)
            this.load()
        
        return _settings
    }
}
